### PR TITLE
Core/formula separation

### PR DESCRIPTION
**Pull request by xu-cheng**
This PR shows all the necessary core code change for core/formula separation for review. **DO NOT Merge IT**. To deliver this change, one should follow below procedures:
* create a new branch in Homebrew namely `tng`(short for the next generation. we could discuss a better name) and push all the changes made in this PR.
* set new branch as default branch.
* merge the first two commits in this PR to master. i.e. tap_migrations hack and new update script.

After these changes, `brew --version` will report `1.0.0`(we could discuss which version should be bumped to) to help us differentiate pre and post migration. 

## How the migration works?

For usres who have old Homebrew installed, running `brew update` will
set motion for following events:
* Old `brew update` logic will update Homebrew to the latest commit in
  master branch, i.e. the second commit.
* When generating update report, `tap_migrations.rb` will be loaded. A
  hack introduced in the first commit will swap the process to run
  `brew update` in second time.
* The new `brew update` logic will set upstream branch to `tng` and
  perform all the necessary migrations.

## What's new file system structure?

* `Library/Formula` and `Library/Aliases` will be replaced as symlinks pointed to the new tap to keep backward compatibility.
* `Library/tap_migrations.rb` will be replaced as a hack explained above.  It has been migrated to `tap_migrations.json` in the `Homebrew/core` tap.
* `Library/formula_renames.rb` will be removed. It has been migrated to `formula_renames.json` in the `Homebrew/core` tap.
* New tap `homebrew/core` will be installed as core tap.
* Shell completion scripts will be moved to `/etc/bash_completion.d/brew` and `/share/zsh/site-functions/_brew`, so they can be picked up automatically by bash and zsh respected. The main thought is to use this migration  as opportunity to avoid git conflict.

## Test 

TL;DR `brew update` works well for all OS X images available on Travis.

I set up https://github.com/xu-cheng/homebrew-test as test repo. It contains two branches master and tng to simulate core repo. And I test it against all OS X images available on Travis. The result is in
https://travis-ci.org/xu-cheng/homebrew-test/builds/113636862. <del>Noted OS X image xcode6.4 fails because it cannot run `brew update` in the first place because the old bug and there isn't anything we can do.</del>(I have found way to fix it)

## Remaining Works (read: to be addressed in future PRs)

- [ ] update all the document in core.
- [ ] set up document for new core tap.
- [ ] set up new CI flow.
  - [x] code changes
  - [x] create `brew legacy-pull` https://gist.github.com/545163571f6c6ae74d6a It can pull homebrew formulae pull request to the new core tap. 
  - [ ] BrewTestBot forks Homebrew/homerbew-core
  - [ ] add Travis CI job to Homebrew/homebrew-core using same .travis.yml from other taps.
  - [ ] add new Jenkins job
- [ ] update brew doctor to include check for new core tap.
- [ ] update install script.
- [ ] update `bash-completion` formula because of shell completion script relocation.

### CODE REVIEW

**Comment by tdsmith** [comment]
Very exciting! Can't wait to play with this. Right now my only bikesheddy comment is that `formulas` might be a more intuitive name for the formulas repo than `core`. Enthusiastic :+1: for version 1.0—or "1.0.0"? Is this a good opportunity to start doing versioned core releases -- i.e. should update jump to the latest tag on the tng branch? Don't think it's strictly necessary to figure that out before we switch this on. If there's any API documentation work left to do before we're confident declaring v1.0, let me know, I'm willing to help.

**Comment by mikemcquaid** [comment]
> Right now my only bikesheddy comment is that formulas might be a more intuitive name for the formulas repo than core

`core` was my idea because this won't be user-exposed and (pretty much) every tap contains formulae.

> Enthusiastic  for version 1.0—or "1.0.0"?

1.0.0 from me too :+1: 

> Is this a good opportunity to start doing versioned core releases -- i.e. should update jump to the latest tag on the tng branch? Don't think it's strictly necessary to figure that out before we switch this on. If there's any API documentation work left to do before we're confident declaring v1.0, let me know, I'm willing to help.

I think using tagged releases like this should maybe be a requirement of a 1.0.0 because it means we can actual use semantic versioning sensibly. I also think we should consider (not for 1.0.0) fixing and supporting `brew tests` on Linux and shipping a RubyGem with Homebrew's core code as an official Ruby API. We'll also probably want to do things like test formulae work with `readall` on both the current stable version and current tip of the `tng` branch (or whatever it ends up being).

**Comment by sjackman** [comment]
Very exciting! Thanks for getting this going, Xu!

On the bikeshed argument, I prefer `core`.

Currently `git` is not a dependency of Homebrew/Linuxbrew. `brew update` can `brew install git` if necessary. Will that still be the case with 1.0.0?

> I also think we should consider (not for 1.0.0) fixing and supporting `brew tests` on Linux

`brew tests` succeeds on Linuxbrew. I'm sure that it should be possible to get that going in Homebrew as well. What's your motivation, Mike?

**Comment by mikemcquaid** [comment]
> brew tests succeeds on Linuxbrew. I'm sure that it should be possible to get that going in Homebrew as well. What's your motivation, Mike?

You'll probably be pleased to hear that my view is that we should make the effort for the core code to support (and have CI for) Linux once we've separated core from the formulae.

**Comment by xu-cheng** [comment]
> Enthusiastic for version 1.0—or "1.0.0"?

It reports 1.0.0 now.

> If there's any API documentation work left 

I haven't touch API yet. I think it may be better to be handled after this PR is shipped. That to be said, there will be lots of changes in FAQ/PR workflow for Formula Author etc. 

> Currently git is not a dependency of Homebrew/Linuxbrew. brew update can brew install git if necessary. Will that still be the case with 1.0.0?

Yes, it's still the same. And this could cause problem if the core tap is not installed, which means git.rb is not available. On homebrew we can always workaround it by using git from Xcode. I don't know how to handle it on Linuxbrew land.

> You'll probably be pleased to hear that my view is that we should make the effort for the core code to support (and have CI for) Linux once we've separated core from the formulae.

Agreed, we should try best to move all os related codes to `Library/os` folder.

**Comment by xu-cheng** [comment]
@sjackman Could you give me the write permission for Linuxbrew org, so I could create `Linuxbrew/homebrew-core` repo similar to what we have in Homebrew? Thanks.

**Comment by mikemcquaid** [comment]
> @sjackman Could you give me the write permission for Linuxbrew org, so I could create Linuxbrew/homebrew-core repo similar to what we have in Homebrew? Thanks.

Obviously it's up to you but I think this is a really good chance for Linuxbrew to specifically only start adding formulae you know work on Linux rather than everything in the Linuxbrew repository.

**Comment by xu-cheng** [comment]
> Obviously it's up to you but I think this is a really good chance for Linuxbrew to specifically only start adding formulae you know work on Linux rather than everything in the Linuxbrew repository.

I'm fine either way. But I think it may be better to do the subtraction instead of addition for compatibility purpose. We could/should discuss this in Linuxbrew land, so it wouldn't create noise for this PR.

**Comment by sjackman** [comment]
> You'll probably be pleased to hear that my view is that we should make the effort for the core code to support (and have CI for) Linux once we've separated core from the formulae.

That's exciting! If you ever peruse a diff of `homebrew/master...linuxbrew/master` I'm happy to comment/clarify under what circumstances particular patches were needed.

> Obviously it's up to you but I think this is a really good chance for Linuxbrew to specifically only start adding formulae you know work on Linux rather than everything in the Linuxbrew repository.

Removing untested formula from Linuxbrew would be a barrier/discouragement for people that need to contribute a one-line fix to get the formula working on Linux, which often is all that's necessary. I'd prefer to maintain a white/grey/black/red list of "known to work", "unknown", "known not to work currently", and "will never ever work, definitely requires Mac OS". Installing grey formulas would give a warning. Installing black formulas would give an error, with a request to fix it and submit a PR. Installing red formula would give an error.

I've been working recently on getting AWS EC2 and Docker Cloud set up for Linuxbrew so that I can get proper CI and bottling set up for Linuxbrew. Bottles work now, but I build them manually on a Docker box and upload them to Bintray, and only for a select few core dependencies. Once the build box is up and running, I'd like to try the grand build of all formula on Linux.

**Comment by mikemcquaid** [comment]
> Removing untested formula from Linuxbrew would be a barrier/discouragement for people that need to contribute a one-line fix to get the formula working on Linux, which often is all that's necessary, and also complicate the git history and merging. I'd prefer to maintain a white/grey/black/red list of "known to work", "unknown", "known not to work currently", and "will never ever work, definitely requires Mac OS". Installing grey formulas would give a warning. Installing black formulas would give an error, with a request to fix it and submit a PR. Installing red formula would give an error.

I don't want to force you to adopt a particular approach but this approach is guaranteed to make Linuxbrew remain an unreliable, low-quality project. You could fork Homebrew's tap and work on tooling to easily make these contributions but I'd really, really strongly advise you to make the most of this opportunity to only adopt those white formulae by default (perhaps offering all Homebrew formulae as an optional tap) to start to be able to move Linuxbrew towards having only formulae known to work.

**Comment by apjanke** [comment]
Yay!

>> Right now my only bikesheddy comment is that formulas might be a more intuitive name for the formulas repo than core
>
> `core` was my idea because this won't be user-exposed and (pretty much) every tap contains formulae.

What name should we use for the system that's left in the homebrew/homebrew repo—the actual package manager program, as opposed to the formulae? It seems like we've been using "core" to refer to that to distinguish it from the formulae, e.g. in the Pull Request template. It's not end-user exposed for people just using Homebrew to install things, but it is contributor exposed, and that's still a lot of people and discussion.

**Comment by apjanke** [comment]
I'm kind of partial to using "core" to mean the package manager implementation (in `homebrew/homebrew`) and something like "base" for the set of formulae in the main/default tap. It'd be analogous to the "base system" of FreeBSD: the set of packages which everyone has available and is tested together as a set. (Though we don't actually maintain the packages; just their installation process.)

**Comment by sjackman** [comment]
That's a good point, Andrew. I have also been using Core to mean the package manager and not the formulas. For what it's worth Debian also uses Base to mean the core set of installed packages, although for Debian, Base is a subset of Main, and Main is more like all the formula available in Homebrew.

**Comment by mikemcquaid** [comment]
> I'm kind of partial to using "core" to mean the package manager implementation (in homebrew/homebrew) and something like "base" for the set of formulae in the main/default tap. It'd be analogous to the "base system" of FreeBSD: the set of packages which everyone has available and is tested together as a set. (Though we don't actually maintain the packages; just their installation process.)

I like the idea of the actual package manager being `brew` (as opposed to `Homebrew` or `Linuxbrew`).

**Comment by apjanke** [comment]
> I like the idea of the actual package manager being `brew` (as opposed to `Homebrew` or `Linuxbrew`).

Me too. That seems like a workable terminology distinction.

**Comment by xu-cheng** [comment]
OK, we have several major issues need to be sorted out first:

###  How do we do the migration?

Currently there are two approaches:
* By switching users to a new branch, i.e. how is done in this PR.
* By creating new core code repository. Presumably we could call the new repo as `Homebrew/brew`. (side question: should we just copy the code or preserve all git history if we choose this approach?)

The second approach is clearly more clean then first one. Also it means we could continue use `master` as the default branch. But it means we would lost all the stars/forks in this repo. And more importantly, since it is not same as renaming the repo, we won't have the url redirects. This means there will be users end up with the wrong repo as we cannot change all the links either in the old document or blogs/twitters etc around Internet.

### What's the name of new core/base/default tap we should use?

These are all the candidates so far:
* `homebrew-core`
* `homebrew-default`
* `homebrew-base`
* `homebrew-formulae`

Both `default` and `base` seem to be easier to understand. But we have use `Formula#core_formula?` as API name for years. Maybe we should consider to rename the API (and set up alias for backward compatibility purpose). That to be said, I'm considering to rename `CoreFormulaRepository` in the future. The name is just too long and too tedious. We could name it to `CoreTap` as the original PR or `DefaultTap`.

### What version number should we report after core/formula separation?

Initially `1.0.0` seems to be a good idea. But due to the possible API changes we are talking about in the above, maybe we should report `0.9.6` or something. `1.0.0` feels like a stable release to me.

**Comment by sjackman** [comment]
> side question: should we just copy the code or preserve all git history if we choose this approach?

:+1+ for preserving the git history.

**Comment by chdiza** [comment]
> side question: should we just copy the code or preserve all git history if we choose this approach?

I would hope that not ALL of the git history gets preserved.  Surely we don't need to cart around stuff from 2009-2011.  And I would hope that a repo devoted just to the package manager internals wouldn't have a gigantic history of all the formula files in it.  Especially if the separate repo of formula files is going to have a duplicate of that history.

**Comment by mikemcquaid** [comment]
> I would hope that not ALL of the git history gets preserved. Surely we don't need to cart around stuff from 2009-2011. And I would hope that a repo devoted just to the package manager internals wouldn't have a gigantic history of all the formula files in it.

We'll filter the history so that only formulae changes are in the formulae repo and core changes in the core/brew repository.

**Comment by mikemcquaid** [comment]
> By switching users to a new branch, i.e. how is done in this PR.
> By creating new core code repository. Presumably we could call the new repo as Homebrew/brew. (side question: should we just copy the code or preserve all git history if we choose this approach?)
> The second approach is clearly more clean then first one. Also it means we could continue use master as the default branch. But it means we would lost all the stars/forks in this repo. And more importantly, since it is not same as renaming the repo, we won't have the url redirects. This means there will be users end up with the wrong repo as we cannot change all the links either in the old document or blogs/twitters etc around Internet.

We (well, GitHub staff e.g. me) can manually set up the repository redirects for this case. We'd lose stars/forks but I think the cleanliness of a new repo may make it worthwhile.

I also think this might be a chance to start optionally using the "merge pull request" button on one/both repositories.

> What's the name of new core/base/default tap we should use?
> 
> These are all the candidates so far:
> 
> homebrew-core
> homebrew-default
> homebrew-base
> homebrew-formulae
> Both default and base seem to be easier to understand. But we have use Formula#core_formula? as API name for years. Maybe we should consider to rename the API (and set up alias for backward compatibility purpose). That to be said, I'm considering to rename CoreFormulaRepository in the future. The name is just too long and too tedious. We could name it to CoreTap as the original PR or DefaultTap.
> 
> What version number should we report after core/formula separation?
> 
> Initially 1.0.0 seems to be a good idea. But due to the possible API changes we are talking about in the above, maybe we should report 0.9.6 or something. 1.0.0 feels like a stable release in terms of internal API to me.

**Comment by mikemcquaid** [comment]
> By creating new core code repository. Presumably we could call the new repo as Homebrew/brew. (side question: should we just copy the code or preserve all git history if we choose this approach?)
> The second approach is clearly more clean then first one. Also it means we could continue use master as the default branch. But it means we would lost all the stars/forks in this repo. And more importantly, since it is not same as renaming the repo, we won't have the url redirects. This means there will be users end up with the wrong repo as we cannot change all the links either in the old document or blogs/twitters etc around Internet.

I think I prefer creating a new core code repository over a new branch but don't have really strong views. It would be good to benchmark e.g. cloning/fetching performance to see how they are affected.

We (well, GitHub staff e.g. me) can manually set up the repository redirects for this case. We'd lose stars/forks but I think the cleanliness of a new repo may make it worthwhile.

I also think this might be a chance to start optionally using the "merge pull request" button on one/both repositories.

> homebrew-core

This is my favourite.

> homebrew-default
> homebrew-base

I'd be fine with either of these.

> homebrew-formulae

I find this one a little confusing just because (almost) all taps include formulae.

> Both default and base seem to be easier to understand. But we have use Formula#core_formula? as API name for years. Maybe we should consider to rename the API (and set up alias for backward compatibility purpose). That to be said, I'm considering to rename CoreFormulaRepository in the future. The name is just too long and too tedious. We could name it to CoreTap as the original PR or DefaultTap.

Yeh, I'm fine with changing internal APIs and 

> What version number should we report after core/formula separation?
> 
> Initially 1.0.0 seems to be a good idea. But due to the possible API changes we are talking about in the above, maybe we should report 0.9.6 or something. 1.0.0 feels like a stable release in terms of internal API to me.

Yeh, I think I'd like to make this 0.9.6 (or even 0.9.9) with the intention of making a 1.0.0 once we have an API we're happy being stable, green and automatic CI for Linux and perhaps a RubyGem release. I don't think any of that should be too time consuming.

**Comment by sjackman** [comment]
> You'll probably be pleased to hear that my view is that we should make the effort for the core code to support (and have CI for) Linux once we've separated core from the formulae.

For my own curiosity, why the change of mind over Linux, Mike?

**Comment by xu-cheng** [comment]
Since we are talking about Linuxbrew any way, and I like we can support it on same core repo. But I want to strength that we should not achieve this by just merging the diff between current Linuxbrew with Homebrew  core code. We need a proper OS abstraction for this. i.e. we should put all Linux related code in `os/linux.rb` instead of using `OS.mac?/linux?` everywhere.

**Comment by sjackman** [comment]
Oh, yes! Absolutely. The Linuxbrew patches are written to keep the diff small, with the understanding that they wouldn't be merged upstream, to reduce the chance of merge conflicts, and to make the burden of merging the two repositories every week manageable. I think the diff between Homebrew and Linuxbrew will be useful in identifying areas where I ran into trouble, and how I solved that trouble, but then can be reimplemented in a tidy manner.

**Comment by mikemcquaid** [comment]
> For my own curiosity, why the change of mind over Linux, Mike?

Basically because it'll be relatively easy to get Linux support working and keep it working for just the core through CI. Also because I'd like to see Linuxbrew or an equivalent project which maintains the whitelist approach I detailed above and this would allow either Linuxbrew to do that or another project to compete with Linuxbrew on that approach while sharing the same code code.

**Comment by sjackman** [comment]
My idea for the white list is to annotate the formulas like so:

Status | Ruby
---|---
Built on CI | `bottle do sha256 "…" => :x86_64_linux end`
Unknown | no annotation
Fails | `fails_with :macos` (or something like this)
MacOS only | `depends_on :macos`

+ Installing a `MacOS only` package will give an appropriate error message.
+ Installing a `Fails` package will give an error message, a link to the GitHub issue, and solicit a pull request to fix.
+ Installing an `Unknown` formula could give an error message by default, which can be downgraded to a warning with a command line option or an environment variable.

The ultimate plan is that there will be no `Unknown` formula. When I merge Homebrew into Linuxbrew first as a pull request, all the modified and new formula will be built on CI and will either be bottled or marked as `Fails`.

**Comment by xu-cheng** [comment]
Move this PR to homebrew/brew#1

**Comment by mikemcquaid** [comment]
> When I merge Homebrew into Linuxbrew first as a pull request, all the modified and new formula will be built on CI and will either be bottled or marked as Fails.

What CI are you using for this and what's the build timeout, out of interest?

**Comment by apjanke** [comment]
How about a ["core-formula split" label](https://github.com/Homebrew/homebrew/labels/core-formula%20split) to track related issues & PRs, since there's likely to be several?

**Comment by apjanke** [comment]
Nevermind about the label: I made a tracking issue #50650 instead. It'll avoid label clutter, and collect issues from multiple repos.
